# Data Analysis with Python
## Testcricket_Bowlers_Analysis
<br>Md. Tawhidul Islam Bhuiyan
<br>Bsc 4th year, Rajshahi University of Engineering & Technology.

**Objectives:**
- Display the first 10 rows of the dataframe
- Create a markdown cell and explain the meaning of each column
- find the number of rows and columns in the dataframe
- find the data statistics and check for the data types
- Are there any missing values present in the dataset?
- Rename the column names appropriately
- Remove a column from the dataframe

**References:**
<br> Dataset Source: https://stats.espncricinfo.com/ci/content/records/93276.html


### Import required libraries

In [1]:
import numpy as np
import pandas as pd

### Reading xlsx file

In [2]:
df = pd.read_excel("test_cricket.xlsx", engine='openpyxl' , sheet_name='wickets')

### First 10 rows of dataset

In [3]:
display(df.head(10))

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0  M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1           SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2         A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3        JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4         GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
5          SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523   
6            CA Walsh (WI)  1984-2001  132   242  30019  12688   519   
7            DW Steyn (SA)  2004-2019   93   171  18608  10077   439   
8      N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434   
9        HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433   

                   BBI                  BBM    Ave  Econ    SR   5  10  
0  1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1  1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2  1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3  1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4  2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
5  2021-08-15 00:00:00               11/121  27.72  2.92  56.8  18   3  
6  1937-07-01 00:00:00                13/55  24.44  2.53  57.8  22   3  
7  1951-07-01 00:00:00  1960-11-01 00:00:00  22.95  3.24  42.3  26   5  
8  1983-09-01 00:00:00               11/146  29.64  2.78  63.9  23   2  
9                9/127               14/184  28.07  2.80  60.0  34   9

### Checking number of rows and columns in data set

#### The Feature variables(columns) of the test bowlers dataset :
**Player**: The name of the players.
<br>**Span**: The period of career time of the players.
<br>**Mat**: The number of test matches. 
<br>**Inns**: The number of innings.
<br>**Balls**: TThe number of balls bowled. 
<br>**Runs**: The number of runs conceded.
<br>**Wkts**: The number of wickets taken.
<br>**BBI**: stands for Best Bowling in Innings and only gives the score for one innings.
<br>**BBM**: stands for Best Bowling in Match and gives the combined score over 2 innings in one match. 
<br>**Ave**: The average number of runs conceded per wicket.
<br>**Econ**: The average number of runs conceded per over. (Econ = Runs/Overs bowled). 
<br>**SR**: The average number of balls bowled per wicket taken. (SR = Balls/W).
<br>**5**: The number of innings in which the bowler took at least five wickets.
<br>**10**: The number of matches in which the bowler took at least ten wickets.


In [4]:
# number of rows
print("number of rows = ", df.shape[0])

# number of columns
print("number of columns = ", df.shape[1])

number of rows =  79
number of columns =  14


### Features in the dataset and search for missing values and data types

In [5]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     int64  
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(7), object(4)
memory usage: 8.8+ KB


None

##### Observations:

There are
> - 3 float type variables- Bowling average, Economy rate, Strike rate<br>
> - 4 String type variables- Best Bowling in Innings, Best Bowling in Match, Player, Span<br>
> - 7 integer type variables- Match, Innings, Balls, Runs, Wickets, 5 wickets in and innings, 10 wickets in an innings

> From the non-null columns, we can see that there are **no missing value** as of now.<br>

### Staistical Analysis of the dataset

In [6]:
display(df.describe())

Mat        Inns         Balls          Runs        Wkts  \
count   79.000000   79.000000     79.000000     79.000000   79.000000   
mean    80.101266  144.797468  18630.303797   8595.506329  317.101266   
std     28.537692   51.042310   7190.036515   3080.256645  121.731587   
min     37.000000   67.000000   8785.000000   4846.000000  200.000000   
25%     60.500000  110.000000  13580.000000   6456.500000  229.000000   
50%     71.000000  129.000000  16498.000000   7742.000000  266.000000   
75%     93.000000  169.000000  21742.500000   9756.000000  374.500000   
max    166.000000  301.000000  44039.000000  18355.000000  800.000000   

             Ave       Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  79.000000  
mean   27.466456   2.806582  59.187342  16.354430   2.797468  
std     3.657561   0.351666   9.349337   9.642372   3.235935  
min    20.940000   1.980000  41.200000   3.000000   0.000000  
25%    24.425000   2.600000  53.300000   9.500000   1.000000  
50%    28.000000   2.820000  57.400000  14.000000   2.000000  
75%    29.870000   3.080000  63.950000  20.500000   3.500000  
max    34.790000   3.460000  91.900000  67.000000  22.000000

##### Statistic measures
- Count- total numbers of observation
- mean- average of the values
- std- standard deviation
- min- minimum value in the column
- 25%- 1st quartile or 25% of all value is under the certain column value
- 50%- 2nd quartile or 50% of all value is under the certain column value
- 75%- 3rd quartie or 75% of all value is under the certain column value
- max- maximum value in the column

##### Observations
> - On an average a played has played 144 innings
> - Minimum 67 innings were played by a player
> - 25% of players have played 110 innings or less
> - 50% of players have played 129 innings or less
> - 75% of players have played 169 innings or less
> - Maximum 301 innings were played by a player
> - Minimum 67 innings were played by a player

Many other primary insights can be taken from the statistics

### Renaming columns in dataset

In [7]:
# renameing columns in dataframe
df = df.rename(columns={'Span' : 'Carrer_Duration',
                        'Mat' : 'Match',
                        'Inns': 'Innings',
                        'Wkts' : 'Wickets',
                        'BBI' : 'Best_Bowling_Innings',
                        'BBM' : 'Best_Bowling_Match',
                        'Ave' : 'Bowling_Average',
                        'Econ': 'Economy_Rate',
                        'SR' : 'Strike_Rate',
                        5: '5_Wickets',
                        10: '10_Wickets'})
display(df.head())

Player Carrer_Duration  Match  Innings  Balls   Runs  \
0  M Muralitharan (ICC/SL)       1992-2010    133      230  44039  18180   
1           SK Warne (AUS)       1992-2007    145      273  40705  17995   
2         A Kumble (INDIA)       1990-2008    132      236  40850  18355   
3        JM Anderson (ENG)       2003-2021    162      301  34791  16457   
4         GD McGrath (AUS)       1993-2007    124      243  29248  12186   

   Wickets Best_Bowling_Innings   Best_Bowling_Match  Bowling_Average  \
0      800  1951-09-01 00:00:00               16/220            22.72   
1      708  1971-08-01 00:00:00               12/128            25.41   
2      619  1974-10-01 00:00:00               14/149            29.65   
3      617  1942-07-01 00:00:00  1971-11-01 00:00:00            26.67   
4      563  2021-08-24 00:00:00  2021-10-27 00:00:00            21.64   

   Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  
0          2.47         55.0         67          22  
1          2.65         57.4         37          10  
2          2.69         65.9         35           8  
3          2.83         56.3         30           3  
4          2.49         51.9         29           3

### Removing columns from dataset

<br> The data in the column BBI and BBM was bugged and some of the entries got turned into datetime
<br> removing Best_Bowling_Innings and Best_Bowling_Match

In [8]:
df.drop(['Best_Bowling_Innings' , 'Best_Bowling_Match'] , axis=1 , inplace= True )

display(df.head())


Player Carrer_Duration  Match  Innings  Balls   Runs  \
0  M Muralitharan (ICC/SL)       1992-2010    133      230  44039  18180   
1           SK Warne (AUS)       1992-2007    145      273  40705  17995   
2         A Kumble (INDIA)       1990-2008    132      236  40850  18355   
3        JM Anderson (ENG)       2003-2021    162      301  34791  16457   
4         GD McGrath (AUS)       1993-2007    124      243  29248  12186   

   Wickets  Bowling_Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  
0      800            22.72          2.47         55.0         67          22  
1      708            25.41          2.65         57.4         37          10  
2      619            29.65          2.69         65.9         35           8  
3      617            26.67          2.83         56.3         30           3  
4      563            21.64          2.49         51.9         29           3

### Splitting Player and Country from 'Player' Column

In [9]:
df['Player'] = df['Player'].str.replace(")"," ")  
df_player = df['Player'].str.split("(", expand=True)
df = df.drop('Player', axis=1)  
df = pd.concat([df, df_player], axis=1) 
df = df.rename(columns={0: 'Player',1: 'Country'})  
column_sequence = ['Player' , 'Country' , 'Carrer_Duration' , 'Match' , 'Innings', 'Balls', 'Runs' , 'Wickets', 'Bowling_Average' , 'Economy_Rate' , 'Strike_Rate' , '5_Wickets' , '10_Wickets']

df = df[column_sequence]

display(df.head())


<ipython-input-9-cb86bf5a3b72>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['Player'] = df['Player'].str.replace(")"," ")


Player  Country Carrer_Duration  Match  Innings  Balls   Runs  \
0  M Muralitharan   ICC/SL        1992-2010    133      230  44039  18180   
1        SK Warne      AUS        1992-2007    145      273  40705  17995   
2        A Kumble    INDIA        1990-2008    132      236  40850  18355   
3     JM Anderson      ENG        2003-2021    162      301  34791  16457   
4      GD McGrath      AUS        1993-2007    124      243  29248  12186   

   Wickets  Bowling_Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  
0      800            22.72          2.47         55.0         67          22  
1      708            25.41          2.65         57.4         37          10  
2      619            29.65          2.69         65.9         35           8  
3      617            26.67          2.83         56.3         30           3  
4      563            21.64          2.49         51.9         29           3


**Checking how many players played for ICC**

In [10]:
# filter function 
def ICC(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

# using filter function to check
df['Played_for_ICC'] = df['Country'].apply(ICC)

# removing "ICC/" from string of country column
df['Country'] = df['Country'].str.replace("ICC/","", regex=True)
display(df.head(5))

Player Country Carrer_Duration  Match  Innings  Balls   Runs  \
0  M Muralitharan      SL        1992-2010    133      230  44039  18180   
1        SK Warne     AUS        1992-2007    145      273  40705  17995   
2        A Kumble   INDIA        1990-2008    132      236  40850  18355   
3     JM Anderson     ENG        2003-2021    162      301  34791  16457   
4      GD McGrath     AUS        1993-2007    124      243  29248  12186   

   Wickets  Bowling_Average  Economy_Rate  Strike_Rate  5_Wickets  10_Wickets  \
0      800            22.72          2.47         55.0         67          22   
1      708            25.41          2.65         57.4         37          10   
2      619            29.65          2.69         65.9         35           8   
3      617            26.67          2.83         56.3         30           3   
4      563            21.64          2.49         51.9         29           3   

  Played_for_ICC  
0            Yes  
1             No  
2             No  
3             No  
4             No

In [11]:
print(df['Played_for_ICC'].value_counts())

No     74
Yes     5
Name: Played_for_ICC, dtype: int64



**Only 5 players played for ICC**

### Checking how many different countries are present in the dataset

In [12]:
print(df['Country'].value_counts())

AUS         18
ENG         13
INDIA       10
WI           9
SA           8
PAK          7
NZ           7
SL           3
ENG/ICC      2
BDESH        1
ZIM          1
Name: Country, dtype: int64


So 18 Australian and 1 Bangladedshi bowlers present in the dataset.

### Splitting career duration from span

In [13]:
# splitting 'Carrer_Duration' column
df_carrer_duration = df['Carrer_Duration'].str.split("-", expand=True)

# concatenating the new dataframe with the main dataframe
df = pd.concat([df, df_carrer_duration], axis=1)

# renaming new splitted column
df = df.rename(columns={0: "Start_Year",
                        1: "End_Year"})

# removing 'Carrer_Duration' column
df = df.drop("Carrer_Duration", axis=1)

# new column sequence as a list
column_sequence = ['Player', 'Country','Start_Year', 'End_Year', 'Match', 'Innings', 'Balls', 'Runs','Wickets' , 'Bowling_Average','Economy_Rate','Strike_Rate', '5_Wickets', '10_Wickets' , 'Played_for_ICC']   

# implementing the new column sequence
df = df[column_sequence]  

**Changing Start and End Year from object type to integer**

In [14]:
df['End_Year']=pd.to_numeric(df['End_Year'])
df['Start_Year']=pd.to_numeric(df['Start_Year'])

In [15]:
# total years of career of the players
df['Career_Span']=df['End_Year']-df['Start_Year']

# rearranging columns
column_sequence = ['Player', 'Country','Start_Year', 'End_Year', 'Career_Span' , 'Match', 'Innings', 'Balls', 'Runs','Wickets' , 'Bowling_Average','Economy_Rate','Strike_Rate', '5_Wickets', '10_Wickets' , 'Played_for_ICC']   
df = df[column_sequence] 
display(df.head())

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
0  M Muralitharan      SL         1992      2010           18    133      230   
1        SK Warne     AUS         1992      2007           15    145      273   
2        A Kumble   INDIA         1990      2008           18    132      236   
3     JM Anderson     ENG         2003      2021           18    162      301   
4      GD McGrath     AUS         1993      2007           14    124      243   

   Balls   Runs  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  \
0  44039  18180      800            22.72          2.47         55.0   
1  40705  17995      708            25.41          2.65         57.4   
2  40850  18355      619            29.65          2.69         65.9   
3  34791  16457      617            26.67          2.83         56.3   
4  29248  12186      563            21.64          2.49         51.9   

   5_Wickets  10_Wickets Played_for_ICC  
0         67          22            Yes  
1         37          10             No  
2         35           8             No  
3         30           3             No  
4         29           3             No

### Players with Longest to Shortest Carrer Span

In [16]:
display(df.sort_values(by='Career_Span', ascending = False).head(3))

display(df.sort_values(by='Career_Span', ascending = True).head(3))


Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
21    Imran Khan     PAK         1971      1992           21     88      142   
55     GS Sobers      WI         1954      1974           20     93      159   
9   HMRKB Herath      SL         1999      2018           19     93      170   

    Balls   Runs  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  \
21  19458   8258      362            22.81          2.54         53.7   
55  21599   7999      235            34.03          2.22         91.9   
9   25993  12157      433            28.07          2.80         60.0   

    5_Wickets  10_Wickets Played_for_ICC  
21         23           6             No  
55          6           0             No  
9          34           9             No

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
44    GP Swann     ENG         2008      2013            5     60      109   
71    K Rabada      SA         2015      2021            6     47       86   
54  Yasir Shah     PAK         2014      2021            7     45       83   

    Balls  Runs  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  \
44  15349  7642      255            29.96          2.98         60.1   
71   8785  4846      213            22.75          3.30         41.2   
54  13601  7247      235            30.83          3.19         57.8   

    5_Wickets  10_Wickets Played_for_ICC  
44         17           3             No  
71         10           4             No  
54         16           3             No


- Imran Khan has longest carrer span of 21 years
- GP Swann Khan has shortest carrer span of 5 years

### Finding the player that has a lowest economy rate

In [17]:
display(df.sort_values(by='Economy_Rate', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
32  LR Gibbs      WI         1958      1976           18     79      148   

    Balls  Runs  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  \
32  27115  8989      309            29.09          1.98         87.7   

    5_Wickets  10_Wickets Played_for_ICC  
32         18           2             No


- LR Gibbs has the lowest economy rate of 1.98

### Finding the player that has a lowest strike rate

In [18]:
display(df.sort_values(by='Strike_Rate', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
71  K Rabada      SA         2015      2021            6     47       86   

    Balls  Runs  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  \
71   8785  4846      213            22.75           3.3         41.2   

    5_Wickets  10_Wickets Played_for_ICC  
71         10           4             No


- K Rabada has the lowest strike rate of 41.2

### Finding the player that has a lowest bowling average

In [19]:
display(df.sort_values(by='Bowling_Average', ascending = True).head(1))

Player Country  Start_Year  End_Year  Career_Span  Match  Innings  \
19  MD Marshall      WI         1978      1991           13     81      151   

    Balls  Runs  Wickets  Bowling_Average  Economy_Rate  Strike_Rate  \
19  17584  7876      376            20.94          2.68         46.7   

    5_Wickets  10_Wickets Played_for_ICC  
19         22           4             No


- MD Marshall has the lowest bowling average of 20.94